pkg/core/resources/manager/cache.go

```diff
@@ -41,20 +39,34 @@ func NewCachedManager(delegate ReadOnlyResourceManager, expirationTime time.Dura
 		delegate: delegate,
 		cache:    cache.New(expirationTime, time.Duration(int64(float64(expirationTime)*0.9))),
 		metrics:  metric,
+		mutexes:  map[string]*sync.Mutex{},
 	}, nil
 }
 
-func (c cachedManager) Get(ctx context.Context, res model.Resource, fs ...store.GetOptionsFunc) error {
+func (c *cachedManager) Get(ctx context.Context, res model.Resource, fs ...store.GetOptionsFunc) error {
 	opts := store.NewGetOptions(fs...)
 	cacheKey := fmt.Sprintf("GET:%s:%s", res.GetType(), opts.HashCode())
 	obj, found := c.cache.Get(cacheKey)
 	if !found {
-		c.metrics.WithLabelValues("get", string(res.GetType()), "miss").Inc()
-		if err := c.delegate.Get(ctx, res, fs...); err != nil {
-			return err
+		// There might be a situation when cache just expired and there are many concurrent goroutines here.
+		// We should only let one fill the cache and let the rest of them wait for it. Otherwise we will be repeating expensive work.
+		mutex := c.mutexFor(cacheKey)
+		mutex.Lock()
+		obj, found = c.cache.Get(cacheKey)
+		if !found {
+			// After many goroutines are unlocked one by one, only one should execute this branch, the rest should retrieve object from the cache
+			c.metrics.WithLabelValues("get", string(res.GetType()), "miss").Inc()
+			if err := c.delegate.Get(ctx, res, fs...); err != nil {
+				mutex.Unlock()
+				return err
+			}
+			c.cache.SetDefault(cacheKey, res)
 		}
-		c.cache.SetDefault(cacheKey, res)
-	} else {
+		mutex.Unlock()
+		c.cleanMutexFor(cacheKey) // We need to cleanup mutexes from the map, otherwise we can see the memory leak.
+	}
+
+	if found {
 		c.metrics.WithLabelValues("get", string(res.GetType()), "hit").Inc()
```

Comment:
Is it correct to count it as a `hit` if at the first try (line 50) `found` was `false`?

Reply:
introduced hit-wait status